### ISSUE TITLE
[PRO] RedisStorage does quite a bit redis calls that could be reduced by using batch/bulk options

### ISSUE THREAD

**Issue by krapans** [rmosolgo/graphql-ruby#3285]
**DISCLAIMER:** Tried to not expose too much details about PRO version code, so the parts of code that I'm talking about could be a bit vague and marked with filename and lines which I would be referring to.

**Is your feature request related to a problem? Please describe.**

Currently graphql ruby redis storage makes quite a lot of redis calls which could be combined in bulk methods, unions or just pipelined.

**Describe the solution you'd like**

* **subscriptions/redis_storage.rb#write_subscription**

*Lines: 47-55*

Currently every event is processed individually. This creates multiple calls to redis. This could be fixed by using calculating all topics in memory and adding only once to set. (line 48) 

With multiple set add calls could be improved by just wrapping into pipelined which would allow not to wait for first response or creating a custom Lua script, which would deliver all set add values to redis server in one call, thus calls towards redis would be decreased.

* **subscriptions/redis_storage.rb#read_subscription**

Currently it accepts only single subscription id, though this could be bulk method by default, if we would just accept list of arguments `method_name(*sub_id)` as redis get for fetching subscriptions supports bulk mode.

So this way, if someone is using this part of API in their application, they have out of the box API which allows handling more than one subscription with one redis call.

* **subscriptions/redis_storage.rb#each_subscription_id**

This could be modified to allow read subscription ids from multiple topics at the same time, and it could be archived by redis sunion which could be used to collect information from redis with one call versus for each topic ping redis separately.

* Non-pro version **graphql/subscriptions.rb#trigger**

Currently trigger supports only one event per call, though we could introduce trigger_bulk action which would create multiple subscription events and pass it towards `execute_all` method.

This would decrease quite a bit the amount of roundtrips that you need to perform towards redis if you want to push trigger lets say to 5 specific user ids.

This would **graphql/subscriptions.rb#execute_all** require additional changes to have options to handle multiple event updates which would combine with `each_subscription_id` method changes thus decreasing redis calls overall in every action.

**Describe alternatives you've considered**

Making these changes as patches locally, though that ruins any upcoming upgrade tree, as we would need to review each changes by hand and basically cherry pick those changes which would be related to RedisStorage.

**Additional context**

Always open for any ideas how this could be solved differently by decreasing roundtrips towards redis which then overall improved performance of application.

**Comment by rmosolgo** [comment]
Hey, thanks for the great write-up! Yes, I definitely agree that there's a lot of low-hanging fruit there regarding Redis round trips. I'll take a look at some improvements, keeping your suggestions in mind, and follow up here soon.

**Comment by rmosolgo** [comment]
I just released graphql-pro 1.17.1 which includes some of this work. A lot of the backend calls were consolidated into Lua scripts, which should reduce the round-trip overhead. I still plan to investigate some of the structural improvements you mentioned above, but I thought I'd share my progress in case you want to check it out so far 👍

**Comment by krapans** [comment]
Overall I would say an amazing job, as you addressed a lot of places in scope of one update, thank you for that.

Only thing that came to my mind is that after all other spots are cleaned up, it would be great to think about the phase “Performance archived / Clean up and structurize code”. Because now I would say you have decreased the majority of calls and the performance goal is definitely on the way to be completed. After this, it would be great to sit back and look at overall redis storage design to keep performance, but split/refactor to add more.

One example/pattern, which is already used in other spots in gem, abstracting Lua in separate classes so it shows clear and small public API and all details are inside. Again, not for all cases we can apply the same way to solve problems, though just as an idea for further improvements.


Thank you once more for addressing this issue so quickly and efficiently.

**Comment by bdurand** [comment]
I've also been looking into the Redis usage in Pro based on performance reports. I had been playing around with moving calls to use the `multi` method as well as a Lua script for deletions in version 1.16.0.

I just took a look at the changes in 1.17.3 and I think there are still a couple of issues. The first is that the calls to evaluate the Lua scripts are using `eval` which requires the script to be passed over the wire on every call. Using `evalsha` with a rescue clause is more efficient.

```ruby
# Evaluate a lua script on redis.
def eval_script(script, *args)
  script_sha = Digest::SHA1.hexdigest(script)
  begin
    @redis.evalsha(script_sha, [], args)
  rescue Redis::CommandError => e
    if e.message.include?("NOSCRIPT")
      script_sha = @redis.script(:load, script)
      retry
    else
      raise e
    end
  end
end
```

For writing new subscriptions, I was able to make use of the Redis `multi` call to make the operation atomic, but keeping the logic in the Ruby code to make it easier to understand and keeping the CPU load on the app server rather than adding load onto the Redis server.

For iterating over subscription ids, I was also able to eliminate one of the calls to `smembers` which is where our Redis server is reporting most of the time is being spent. I was also able to eliminate the extra calls for garbage collecting the unused keys while still keeping the logic in Ruby.

I think there could also be some other concerns if there are a lot of subscriptions when using the Lua scripts to iterate since each script will be run atomically on the Redis server and block other commands.

I can post my code here or, since this is Pro code, send it directly if you'd like.

**Comment by rmosolgo** [comment]
Thanks for those suggestions, @bdurand! I guess I underestimated the overhead of `.eval` -- I've always used it with static strings, figuring that Redis's deduping would address that problem. I love the approach with `rescue` -- I'll give it a shot. 

I'd love to see which calls you were able to eliminate. Sharing those diffs here or by email would both be fine. I saw some opportunities for structural improvement when I was looking at this before, but haven't had a chance to address it yet!

**Comment by bdurand** [comment]
@rmosolgo Here's the diff against RedisStorage version 1.16.0: https://gist.github.com/bdurand/c24f150e6a82425e1f75671b2118c63b

**Comment by rmosolgo** [comment]
Thanks again for the suggestions above. I just released graphql-pro 1.17.5 with a few changes: 

- Uses `MULTI` when applicable, for simpler implementation, like you mentioned. 
- Adds `EVALSHA` usage, as mentioned above, to avoid running duplicate scripts.


I'm going to look into your `BATCH_SIZE`, too. I could imagine the system having a hard time with a _large_ number of subscriptions for a single event (thousands?) but I haven't heard any instances of it yet. How did you pick `BATCH_SIZE = 24`?  (In any case, I'm considering adding a `batch_size:` option for folks who want to tune that behavior.)

**Comment by bdurand** [comment]
Yeah, the batch size is too low in my code. I'd set it at 24 just so I could see it actually working without having to create 100's of subscriptions and forgot to readjust it.

**Comment by rmosolgo** [comment]
Thanks again for the report and suggestions here. I don't have further plans to improve these calls, so I'll close the issue. If you have problems with the current code, or further suggestions, please open a new issue!
